include/mbedtls/asn1write.h

```diff
@@ -153,34 +153,75 @@ int mbedtls_asn1_write_bool( unsigned char **p, unsigned char *start, int boolea
 int mbedtls_asn1_write_int( unsigned char **p, unsigned char *start, int val );
 
 /**
- * \brief           Write a printable string tag (MBEDTLS_ASN1_PRINTABLE_STRING) and
- *                  value in ASN.1 format
+ * \brief           Write a string in ASN.1 format using a specific
+ *                  string encoding tag.
  *                  Note: function works backwards in data buffer
  *
- * \param p         reference to current position pointer
- * \param start     start of the buffer (for bounds-checking)
- * \param text      the text to write
- * \param text_len  length of the text
+ * \param p         The reference to the current position pointer.
+ * \param start     The start of the buffer (for bounds-checking).
+ * \param tag       The string encoding tag to write, e.g.
+ *                  #MBEDTLS_ASN1_UTF8_STRING.
+ * \param text      The string to write.
+ * \param text_len  The length of \p text in bytes (which might
+ *                  be strictly larger than the number of characters).
  *
- * \return          the length written or a negative error code
+ * \return          The number of bytes written to \p p on success.
+ * \return          A negative error code on failure.
  */
-int mbedtls_asn1_write_printable_string( unsigned char **p, unsigned char *start,
-                                 const char *text, size_t text_len );
+int mbedtls_asn1_write_tagged_string( unsigned char **p, unsigned char *start,
```

Comment:
@sbutcher-arm I think it is worth adding the flexibility here because there are string types other than the ones for which we have a dedicated `mbedtls_asn1_write_xxx_string()` function (see https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/asn1.h#L81 onwards).

Reply:
I agree with Hanno, I think this function is useful as part of the public API. I acknowledge the potential for misuse, but OTOH the entire asn1write module is hardly high-level and has a pretty strong misuse potential already, so I don't think adding this function is a big problem in that respect. Hopefully most users never feel a need to be using asn1write directly :)

Reply:
Okay, I'm persuaded. Thinking about it, I don't think we need to deprecate the `mbedtls_asn1_write_xxx_string()` functions, as they're small, and shouldn't have much overhead in the overall image. Removing them would reduce the overall size by only a few bytes, which at this time I don't think is worth the trouble.

As for the interface, they're undoubtedly easier to use, so I don't see a problem in keep all the functions. 

So for me, that resolves the issue. 

Thanks @mpg and @hanno-arm for the feedback!

Reply:
Marked as 'approved for design'.